Switch to using globalconfig instead of ruleset for analyzer configuration

stephentoub opened the pull request:
cc: @ViktorHofer

stephentoub committed:
Switch to using globalconfig instead of ruleset for analyzer configuration

ViktorHofer commented:
You are suffixing the src and test editorconfig files with .globalconfig so that the default global_level of 100 is chosen, right? I checked the other ones, and they either have -2 or 0 as the global_level so ours should win over the others. Did you try conflicting severity levels with the editorconfig from the platform analyzer out to test this?

ViktorHofer commented on .editorconfig:
```diff
@@ -157,12 +154,6 @@ csharp_space_between_method_declaration_parameter_list_parentheses = false
-dotnet_code_quality.CA1052.api_surface = private, internal
-dotnet_code_quality.CA1802.api_surface = private, internal
-dotnet_code_quality.CA1822.api_surface = private, internal
-dotnet_code_quality.CA2208.api_surface = public
```
Why are these 5 entries removed?

stephentoub replied:
Because they don't belong here.  Some of these rules are disabled, so specifying what surface area we want them applied to isn't useless.  The rest were moved over to the globalconfig next to where we set the severity level.

stephentoub commented:
> Did you try conflicting severity levels with the editorconfig from the platform analyzer out to test this?

Can you give me an example of what I should try?

ViktorHofer commented:
> Can you give me an example of what I should try?

Set any of those to a different level an make sure that the setting in our .editorconfig wins. I would also test it in both the .editorconfig in the root and in the one under eng for source projects.

```
# CA1418: Use valid platform string
dotnet_diagnostic.CA1418.severity = none

# CA1419: Provide a public parameterless constructor for concrete types derived from 'System.Runtime.InteropServices.SafeHandle'
dotnet_diagnostic.CA1419.severity = none

# CA1727: Use PascalCase for named placeholders
dotnet_diagnostic.CA1727.severity = none

# CA1839: Use 'Environment.ProcessPath'
dotnet_diagnostic.CA1839.severity = none

# CA1840: Use 'Environment.CurrentManagedThreadId'
dotnet_diagnostic.CA1840.severity = none

# CA1841: Prefer Dictionary.Contains methods
dotnet_diagnostic.CA1841.severity = none

# CA1842: Do not use 'WhenAll' with a single task
dotnet_diagnostic.CA1842.severity = none

# CA1843: Do not use 'WaitAll' with a single task
dotnet_diagnostic.CA1843.severity = none

# CA1844: Provide memory-based overrides of async methods when subclassing 'Stream'
dotnet_diagnostic.CA1844.severity = none

# CA1845: Use span-based 'string.Concat'
dotnet_diagnostic.CA1845.severity = none

# CA1846: Prefer 'AsSpan' over 'Substring'
dotnet_diagnostic.CA1846.severity = none

# CA1847: Use char literal for a single character lookup
dotnet_diagnostic.CA1847.severity = none

# CA1848: Use the LoggerMessage delegates
dotnet_diagnostic.CA1848.severity = none

# CA2017: Parameter count mismatch
dotnet_diagnostic.CA2017.severity = none

# CA2250: Use 'ThrowIfCancellationRequested'
dotnet_diagnostic.CA2250.severity = none

# CA2251: Use 'string.Equals'
dotnet_diagnostic.CA2251.severity = none

# CA2252: This API requires opting into preview features
dotnet_diagnostic.CA2252.severity = none

# CA2253: Named placeholders should not be numeric values
dotnet_diagnostic.CA2253.severity = none

# CA2254: Template should be a static expression
dotnet_diagnostic.CA2254.severity = none
```

stephentoub commented:
> Set any of those to a different level an make sure that the setting in our .editorconfig wins

Before putting up the PR I'd tried introducing a violation of CA1845 in one of our libs, and it correctly failed the build on my Windows machine.

Is that sufficient to address your concern, or have I not understood?

ViktorHofer commented:
> Is that sufficient to address your concern, or have I not understood?

That's half of the my as that rule is declared in the global editorconfig for source files. You have validated that our global editorconfig file wins over the one from the platformanalyzer package.

Can you please also introduce a violation for which the rule is specified in the root .editorconfig file so that we know for sure that non global editorconfig files win over global ones?

ViktorHofer commented:
In example by adding `dotnet_diagnostic.CA1845.severity = none` to the non global .editorconfig file in the repo root. That severity should win over the other ones in the global editorconfig files.

stephentoub commented:
> Can you please also introduce a violation for which the rule is specified in the root .editorconfig file so that we know for sure that non global editorconfig files win over global ones?

We shouldn't be adding any analyzer configuration to the root .editorconfig moving forward.

But I can certainly test it.  You want me to add a none setting to a rule we define as warning in the global one, introduce a violation, and make sure there isn't a warning?

ViktorHofer commented:
> But I can certainly test it. You want me to add a none setting to a rule we define as warning in the global one, introduce a violation, and make sure there isn't a warning?

Exactly.

stephentoub commented:
I introduced a violation of 1845 in System.Net.Http and it triggered a warning-as-error.  Then I added:
```
dotnet_diagnostic.CA1845.severity = none
```
to the root .editorconfig, and it silenced the warning.

ViktorHofer commented:
Awesome, thanks for trying out the experience.
